## C6M1L2 Item 08 - Making your connection


## Prerequisites: 

To complete this lab, you need to install Python and MySQL databases. Then install `MySQL Connector/Python` client or API on your Python environment. Follow the instructions in previous videos on “installation and setup” to install the required packages or software.  

To work with MySQL using Python, you must have an authorized user account on the MySQL server.  

 

## Scenario:

Little Lemon is developing a Python-based application that needs to use MySQL in the back end so that they can store day-to-day data on bookings, orders and menus in their database’s respective tables.  They have all the required components installed to store the data. However, they need help to establish a connection between Python and the MySQL server so that they can access and communicate with the database using Python. 

You are tasked to help Little Lemon to accomplish its business goals by establishing a secure connection to the MySQL database using Python.  

Note: Need to install mysql-connector-python package to run your code. 

In [12]:
#!pip install mysql-connector-python
#!pip install mysql-connector
!pip install --upgrade mysql-connector-python 

  Obtaining dependency information for mysql-connector-python from https://files.pythonhosted.org/packages/53/ed/26a4b8cacb8852c6fd97d2d58a7f2591c41989807ea82bd8d9725a4e6937/mysql_connector_python-8.3.0-py2.py3-none-any.whl.metadata
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 557.9/557.9 kB 8.1 MB/s eta 0:00:001 MB/s eta 0:00:01
  Attempting uninstall: mysql-connector-python
    Found existing installation: mysql-connector-python 8.0.31
    Uninstalling mysql-connector-python-8.0.31:
      Successfully uninstalled mysql-connector-python-8.0.31

[notice] A new release of pip is available: 23.2 -> 24.0
[notice] To update, run: pip install --upgrade pip


## Task 1:


Help Little Lemon in establishing a connection to their MySQL database using Python, via MySQL Connector/Python API so that they can access data according to their business needs.  

To achieve this task please fill in the following steps: 

* 1: import the appropriate MySQL connector library using the alias connector 

* 2: Next, create a variable called connection and use it to store an instance of the connection made with the database using the connector module. This module uses a method called connect() and you should provide the relevant connection details.  

 

**TIP:** You need to have an authenticated username and password to establish the connection. If you don’t see any error in the output while establishing the connection, your connection is successfully established. 

In [22]:
# Print a message to guide the user
# Import MySQL Connector/Python 
print("Importing MySQL Connector/Python API")
import mysql.connector as connector
print("MySQL Connector/Python API is imported successfully.\n")

# Establis connection with authorized user/password
print("Establishing a new connection between MySQL and Python.")
#connection=connector.connect(user="your_username",password="your_password")

connection=connector.connect(user="admin-meta",password="admin")
print("A connection between MySQL and Python is successfully established")

Importing MySQL Connector/Python API
MySQL Connector/Python API is imported successfully.

Establishing a new connection between MySQL and Python.


ProgrammingError: Character set 'utf8' unsupported

## Task 2:

Little Lemon is interested in writing clean safe code and wants to ensure that there are no connection issues when connecting with MySQL database. They have heard of Python’s try/except blocks and are interested in incorporating them into their own code.   

* 1: As above import an instance of MySQL connector using the alias connector.  
* 2: Open a try block and place the appropriate connection code there  

```Python 
try: 
```
         // using connector.connect() make your database connection.  

 

* 3: Now fill in the code for when an error occurs. What is an appropriate message to display on the screen? 

 
```Python
except: 
```
        // print a meaningful message if the connection has an issue with the `username` or `password`	 
 

**TIP:** Test your set-up by purposefully entering an incorrect user or password and see if you get the following message:  

```Python
There was a problem connecting to the MySQL server.  
Please check the username or password.  
```


In [23]:
import mysql.connector as connector

try:
    connection=connector.connect(
        user="wrong_user",
        password='wrong_password'
    )
except:
    print("""
    There was a problem connecting to the database.
    Please check your username or the password.
    """)


    There was a problem connecting to the database.
    Please check your username or the password.
    


## Task 3:

Little lemon wants to know the reason for any issues that might occur when connecting to the database. Help them to retrieve the error code along with the error message in the following situations:  

* The database does not exist 
* Wrong Username or the password 

To achieve this please you can use the following steps.  
* 1: First import an instance of MySQL connector using the alias connector  

* 2: Open a try block and place the appropriate connection code there. 

```Python
try: 
```    

        // using connector.connect() make your database connection.  


* 3: Finally, fill in the code for when an error occurs. In the previous task, you used the print statement to output a custom error message. Now, improve your approach by using connector’s error exception in the except block.  

```Python
except connector.?() as er: 
```

        // print(“Error code: “, er.?) 
        // print( (“Error message: “, er.?) 

**TIP:** Use the “Error” class from “mysql.connector” to get the error code and the message. Use try-except from Python for implementation.

In [24]:
import mysql.connector as connector

try:
    connection=connector.connect(
        user="meta",
        password="password",
        database = "no_database")

except connector.Error as er:
    print("Error code:", er.errno)
    print("Error message:", er.msg)

Error code: 1045
Error message: Access denied for user 'meta'@'localhost' (using password: YES)


The connect function throws an error Access denied for user with an error code 1045 if it encounters an unauthenticated user [wrong user and/or password].

In [27]:
# Let's close the cursor and the connection
if connection.is_connected():
    connection.close()
    print("MySQL connection is closed.")
else:
    print("Connection is already closed")

NameError: name 'connection' is not defined

In [26]:
try:
    connection=connector.connect(
        user="amdin-meta",
        password="admin",
        #database = "no_database"
    )

except connector.Error as er:
    print("Error code:", er.errno)
    print("Error message:", er.msg)

Error code: 1045
Error message: Access denied for user 'amdin-meta'@'localhost' (using password: YES)
